src/contourbuilder.rs

```diff
@@ -74,32 +76,37 @@ impl ContourBuilder {
         self
     }
 
-    fn smoooth_linear(&self, ring: &mut Ring, values: &[Float], value: Float) {
+    fn smooth_linear<V: GridValue>(&self, ring: &mut Ring, values: &[V], value: V) -> Result<()> {
         let dx = self.dx;
         let dy = self.dy;
         let len_values = values.len();
 
-        ring.iter_mut()
-            .map(|point| {
-                let x = point.x;
-                let y = point.y;
-                let xt = x.trunc() as usize;
-                let yt = y.trunc() as usize;
-                let mut v0;
-                let ix = yt * dx + xt;
-                if ix < len_values {
-                    let v1 = values[ix];
-                    if x > 0.0 && x < (dx as Float) && (xt as Float - x).abs() < Float::EPSILON {
-                        v0 = values[yt * dx + xt - 1];
-                        point.x = x + (value - v0) / (v1 - v0) - 0.5;
-                    }
-                    if y > 0.0 && y < (dy as Float) && (yt as Float - y).abs() < Float::EPSILON {
-                        v0 = values[(yt - 1) * dx + xt];
-                        point.y = y + (value - v0) / (v1 - v0) - 0.5;
-                    }
+        fn cast<V: GridValue>(num: V) -> Result<Float> {
+            num_traits::cast::<V, Float>(num).ok_or_else(|| new_error(ErrorKind::BadCast))
+        }
```

Comment:
Maybe we could turn this into a macro to save us a few function calls?
Or maybe the compiler is clever enough to inline it anyway.

Reply:
I assumed it would inline, but I'll make it a macro just in case.